### DISCUSSION TITLE
Windows 10/21H1 using an X.509 user certificate

### DISCUSSION THREAD

**Discussion by robertdahlem** [strongswan/strongswan#588] [Q&A]
Hi,

I'm trying to setup a VPN connection from Windows 10/21H1 using an X.509 user certificate to Strongswan 5.7.2 under Debian 10. The server has a Let's Encrypt certificate and its own CA for client certificates. The user certificate is signed by the latter.

My PKI has been constructed like this:

```
pki --gen > caKey.der
pki --self --in caKey.der --dn "C=DE, O=vpn.myserver.com, CN=strongSwan CA" --lifetime 7305 --ca > caCert.der
openssl x509 -inform der -in caCert.der -outform pem -out caCert.pem
pki --gen --outform pem >dahlem.key
pki --issue --in dahlem.key --type priv --cacert caCert.der --cakey caKey.der \
    --dn "C=DE, O=vpn.myserver.com, CN=dahlem" --san dahlem --outform pem >dahlem.pem
openssl pkcs12 -in dahlem.pem -inkey dahlem.key                      -export -out dahlem-without-CA.p12
openssl pkcs12 -in dahlem.pem -inkey dahlem.key -certfile caCert.pem -export -out dahlem-with-CA.p12
```

I imported the client certificate dahlem-with-CA.p12 by double clicking on it as described in [Windows Clients](https://wiki.strongswan.org/projects/strongswan/wiki/WindowsClients#On-the-Windows-Client-2), storing it on the Local Computer Level and letting Windows choose the store automatically.

Using the MMC I see the certificate ends up in the computers Personal store and my own CA in Trusted Root Cert Authorities.

Then I click through:

- Set up a Connection or Network
- Connect to a workplace
- Use my Internet connection (VPN)
- Internet address: vpn.myserver.com
- [ x ] Remember my credentials

Switch to the properties of this connection
- Security
- Type of VPN: IKEv2
- Data encryption: maximum strength
- Use machine certificate

This gives me an event ID 20227 with reason code 13801 ("IKE authentication credentials are unacceptable").
On the server side I get: [log-with-CA.txt](https://github.com/strongswan/strongswan/files/7130574/log-with-CA.txt)

Then I deleted both certificates, imported dahlem-without-CA.p12 and re-created the VPN connection.
This gives me an event ID 20227 with reason code 13806 ("IKE failed to find a valid machine certificate").
On the server side I get: [log-without-CA.txt](https://github.com/strongswan/strongswan/files/7130578/log-without-CA.txt)

Configuration: [swanctl.conf.txt](https://github.com/strongswan/strongswan/files/7130582/swanctl.conf.txt)
Loaded Certificates: [list-certs.txt](https://github.com/strongswan/strongswan/files/7130583/list-certs.txt)

Regards,
Robert

**Comment by tobiasbrunner** [comment]
>     * Use machine certificate

That's incorrect if you want to use user certificates, which uses EAP-TLS. Either store the certificate/key in the _Local Computer_ store, or switch to EAP-TLS. Whether the client requires the CA certificate for the client certificate for either authentication method, I don't know.

**Reply by robertdahlem** [reply]
I finally figured out what those codes mean.

13801 means "there is something wrong with your client certificate". Most of the time this will be because it is not stored in the Local Computer store, just as Tobias wrote. In my case it was because I used EAP instead of machine certificates and/or the certificate was stored in my user account. For `connections.<conn>.remote<suffix>.auth=pubkey` one has to store it in the Local Computer store and tell Windows to use machine certficates.

13806 means "there is something wrong with your server certificate". When using Let's Encrypt this might be because your don't have the root CA (nowadays: ISRG Root X1) and/or R3, the intermediate certificate.

You will get ISRG Root X1 once you visit [https://valid-isrgrootx1.letsencrypt.org](https://valid-isrgrootx1.letsencrypt.org/) with IE or Edge (Chrome will not do the trick), because Windows 10 is lazy loading root certificates. Fresh installs simply don't have X1.

I had to install R3 too, which I got from Let's Encrypt as `/etc/letsencrypt/live/vpn.myserver.com/chain.pem` in the Local Computer store under Intermediate Certification Authorities (it gets installed there if you let Windows do it automatically).

What still puzzles me: why would I need to install an intermediate certificate on the client when the server has it under x509 or x509ca (none works, although it shows up in `swanctl --list-certs` in both cases)? Or the other way round: where would I need to place the intermediate certificate for StrongSwan to send it along with its server certificate? Mentioning it in `connections.<conn>.local<suffix>.certs` doesn't help.

**Reply by tobiasbrunner** [reply]
> why would I need to install an intermediate certificate on the client when the server has it under x509 or x509ca (none works, although it shows up in `swanctl --list-certs` in both cases)?

You don't.

> Or the other way round: where would I need to place the intermediate certificate for StrongSwan to send it along with its server certificate?

Strictly speaking, `x509ca` is the correct location, but as discussed before, the certificate is loaded, used and trusted either way.

> Mentioning it in `connections.<conn>.local<suffix>.certs` doesn't help.

That's definitely wrong as that's for end-entity certificates (read the docs).

Check the log for certificate requests and refer to https://github.com/strongswan/strongswan/discussions/473#discussioncomment-1164690 for the possible issues with Let's Encrypt intermediate CA certificates and, of course, our previous discussion in #567.
